File: python/ch03_agent/agent_ch03/core/permissions.py
"""权限策略领域模型。

Java 对照：这里相当于一个独立的 Policy Service 模块，包含不可变 DTO、
规则对象以及可注入的审批/审计接口。它不直接执行工具，只负责回答“能不能执行”。
"""

from collections.abc import Callable, Sequence
from dataclasses import dataclass
from typing import Literal, Protocol

from .filesystem import WorkspaceWriteBoundary
from .tools import PreparedToolCall, ToolContext, ToolResult, tool_error

PermissionBehavior = Literal["allow", "deny", "ask", "passthrough"]
PERMISSION_BEHAVIORS: tuple[PermissionBehavior, ...] = ("allow", "deny", "ask", "passthrough")


class PermissionContractError(Exception):
    """权限请求、规则或决策违反领域契约。"""


def _is_behavior(value: object) -> bool:
    return value in PERMISSION_BEHAVIORS


@dataclass(frozen=True, slots=True)
class PermissionDecision:
    """一次权限结论：行为、解释原因和决策来源。"""

    behavior: PermissionBehavior  # 四态行为；最终只有 allow/deny 可执行或回填。
    reason: str  # 给用户、模型和审计记录看的非空解释。
    source: str  # 规则、审批器、工作区边界或默认策略的来源名。

    def __post_init__(self) -> None:
        if not _is_behavior(self.behavior):
            raise PermissionContractError("behavior 必须是受支持的权限行为")
        if not isinstance(self.reason, str) or not self.reason.strip():
            raise PermissionContractError("权限决策原因不能为空")
        if not isinstance(self.source, str) or not self.source.strip():
            raise PermissionContractError("权限决策来源不能为空")

    @property
    def is_allowed(self) -> bool:
        """只有最终 allow 才允许工具进入 invoke。"""
        return self.behavior == "allow"

    def to_tool_result(self) -> ToolResult:
        """把最终 deny 转成模型可见错误；ask/passthrough 不能直接转换。"""
        if self.behavior != "deny":
            raise PermissionContractError("只有最终 deny 决策才能转换成工具结果")
        return tool_error("permission_denied", self.reason)


@dataclass(frozen=True, slots=True)
class PermissionRequest:
    """权限策略输入快照，必须包含已经通过工具参数校验的调用。"""

    prepared: PreparedToolCall  # prepare 成功的工具调用，权限层不修复坏参数。
    context: ToolContext  # 工作区和调用身份边界。
    recommendations: tuple[PermissionDecision, ...] = ()  # 上游 hook 提供的候选建议。
    proposed_decision: PermissionDecision | None = None  # 交给审批器确认的 ask 决策。

    def __post_init__(self) -> None:
        if self.prepared.error is not None or self.prepared.definition is None or self.prepared.arguments is None:
            raise PermissionContractError("权限请求必须包含准备完成的工具调用")
        if not isinstance(self.recommendations, tuple) or not all(isinstance(item, PermissionDecision) for item in self.recommendations):
            raise PermissionContractError("recommendations 必须全部是 PermissionDecision")
        if self.proposed_decision is not None and (not isinstance(self.proposed_decision, PermissionDecision) or self.proposed_decision.behavior != "ask"):
            raise PermissionContractError("proposed_decision 必须是 ask 决策")


PermissionMatcher = Callable[[PermissionRequest], bool]


@dataclass(frozen=True, slots=True)
class PermissionRule:
    """把匹配条件、行为和解释原因绑定在一起的不可变规则。"""

    name: str  # 审计和决策 source 使用的稳定规则名。
    behavior: PermissionBehavior  # 匹配成功时提出的候选行为。
    reason: str  # 规则为什么提出这个行为。
    matches: PermissionMatcher  # 类似 Java Predicate<PermissionRequest>。

    def __post_init__(self) -> None:
        if not self.name.strip() or not self.reason.strip() or not callable(self.matches) or not _is_behavior(self.behavior):
            raise PermissionContractError("权限规则字段不合法")

    def evaluate(self, request: PermissionRequest) -> PermissionDecision | None:
        """匹配成功返回带来源的候选决策，不匹配返回 None。"""
        if not self.matches(request):
            return None
        return PermissionDecision(self.behavior, self.reason, self.name)


class ApprovalProvider(Protocol):
    """审批边界，类似 Java 的 ApprovalProvider interface。"""

    def decide(self, request: PermissionRequest) -> PermissionDecision:
        """把 ask 收敛成明确 allow 或 deny。"""


class AuditSink(Protocol):
    """审计边界，最终决策记录失败时阻止工具执行。"""

    def record(self, request: PermissionRequest, decision: PermissionDecision) -> None:
        """记录一次最终权限决定。"""


def _strongest(decisions: Sequence[PermissionDecision]) -> PermissionDecision:
    """冲突时按 deny > ask > allow 选择最保守候选。"""
    for behavior in ("deny", "ask", "allow"):
        for decision in decisions:
            if decision.behavior == behavior:
                return decision
    return PermissionDecision("passthrough", "没有权限参与方阻止请求", "default")


def _approval_denial(proposed: PermissionDecision) -> PermissionDecision:
    return PermissionDecision("deny", f"审批没有明确同意: {proposed.reason}", "approval")


class PermissionPolicy:
    """在工具 handler 之前合并工作区边界、默认策略、规则、审批和审计。"""

    def __init__(
        self,
        rules: Sequence[PermissionRule] = (),
        approval: ApprovalProvider | None = None,
        audit: AuditSink | None = None,
        write_boundary: WorkspaceWriteBoundary | None = None,
    ) -> None:
        if not all(isinstance(rule, PermissionRule) for rule in rules):
            raise PermissionContractError("rules 必须全部是 PermissionRule")
        self._rules = tuple(rules)  # 不可变规则快照，类似 Java List.copyOf。
        self._approval = approval  # ask 的外部审批器。
        self._audit = audit  # 最终决定的审计接收器。
        self._write_boundary = write_boundary  # 写路径真实边界检查。

    def decide(self, request: PermissionRequest) -> PermissionDecision:
        """计算唯一最终决定；审计失败会抛出并阻止副作用。"""
        candidates: list[PermissionDecision] = []
        boundary = self._workspace_boundary_decision(request)
        if boundary is not None:
            candidates.append(boundary)
        definition = request.prepared.definition
        if definition is None:
            raise PermissionContractError("权限请求丢失工具定义")
        if definition.effect == "execute":
            candidates.append(PermissionDecision("ask", "执行 PowerShell 命令需要审批", "shell-default"))
        candidates.extend(request.recommendations)
        for rule in self._rules:
            try:
                decision = rule.evaluate(request)
            except Exception:  # noqa: BLE001
                decision = PermissionDecision("deny", f"权限规则执行失败: {rule.name}", rule.name)
            if decision is not None:
                candidates.append(decision)
        proposed = _strongest(candidates)
        if proposed.behavior == "ask":
            final = self._resolve_approval(request, proposed)
        elif proposed.behavior == "passthrough":
            final = PermissionDecision("allow", "没有权限规则阻止请求", "default")
        else:
            final = proposed
        if self._audit is not None:
            self._audit.record(request, final)
        return final

    def _resolve_approval(self, request: PermissionRequest, proposed: PermissionDecision) -> PermissionDecision:
        """审批缺失、异常、无效或返回中间态时统一拒绝。"""
        if self._approval is None:
            return _approval_denial(proposed)
        approval_request = PermissionRequest(request.prepared, request.context, request.recommendations, proposed)
        try:
            decision = self._approval.decide(approval_request)
        except Exception:  # noqa: BLE001
            return PermissionDecision("deny", "审批器发生异常，已拒绝请求", "approval")
        if not isinstance(decision, PermissionDecision):
            return PermissionDecision("deny", "审批器返回了无效决定", "approval")
        if decision.behavior in {"allow", "deny"}:
            return decision
        return _approval_denial(proposed)

    def _workspace_boundary_decision(self, request: PermissionRequest) -> PermissionDecision | None:
        """写工具先检查相对路径；该 deny 不能被 allow、ask 或审批覆盖。"""
        definition = request.prepared.definition
        arguments = request.prepared.arguments
        if definition is None or arguments is None or definition.effect != "write":
            return None
        raw_path = arguments.get("path") if isinstance(arguments, dict) else None
        if raw_path is None or self._write_boundary is None:
            return None
        if not isinstance(raw_path, str):
            return PermissionDecision("deny", "写入路径无效", "workspace-boundary")
        try:
            allowed = self._write_boundary.is_path_within_workspace(request.context.workspace, raw_path)
            if not isinstance(allowed, bool):
                raise PermissionContractError("写路径边界必须返回 bool")
        except Exception:  # noqa: BLE001
            return PermissionDecision("deny", "无法安全解析写入路径", "workspace-boundary")
        if not allowed:
            return PermissionDecision("deny", "禁止写入工作区之外", "workspace-boundary")
        return None

File: typescript/chapters/ch18/src/core/permissions.ts
// 权限策略在工具执行前集中裁决，并将最终决策写入审计边界；P16 的 plan gate 以不可变规则组合进入队友策略。
import type { WorkspaceWriteBoundary } from "./filesystem.js";
import type { PreparedToolCall, ToolContext, ToolResult } from "./tools.js";
import { toolError } from "./tools.js";

export const PERMISSION_BEHAVIORS = Object.freeze(["allow", "deny", "ask", "passthrough"] as const);
export type PermissionBehavior = (typeof PERMISSION_BEHAVIORS)[number];

export class PermissionContractError extends Error {}

export function isPermissionBehavior(value: unknown): value is PermissionBehavior {
  return PERMISSION_BEHAVIORS.some((behavior) => behavior === value);
}

export class PermissionDecision {
  // 最终行为；deny/ask/allow/passthrough 的优先级由策略合并器统一解释。
  readonly behavior: PermissionBehavior;
  // 面向用户和审计的稳定理由，不承载异常堆栈。
  readonly reason: string;
  // 标识决定来源，便于区分系统边界、规则、审批和默认路径。
  readonly source: string;

  constructor(behavior: PermissionBehavior, reason: string, source: string) {
    if (!isPermissionBehavior(behavior)) {
      throw new PermissionContractError("behavior must be a PermissionBehavior");
    }
    if (typeof reason !== "string" || reason.trim().length === 0) {
      throw new PermissionContractError("permission decision reason must not be empty");
    }
    if (typeof source !== "string" || source.trim().length === 0) {
      throw new PermissionContractError("permission decision source must not be empty");
    }
    this.behavior = behavior;
    this.reason = reason;
    this.source = source;
    Object.freeze(this);
  }

  get isAllowed(): boolean {
    return this.behavior === "allow";
  }

  toToolResult(): ToolResult {
    if (this.behavior !== "deny") {
      throw new PermissionContractError("only a final deny decision can become a tool result");
    }
    return toolError("permission_denied", this.reason);
  }
}

export interface PermissionRequestOptions {
  // 已完成 schema 校验且被冻结的调用快照，权限层不能重新解析原始 JSON。
  readonly prepared: PreparedToolCall;
  // workspace 和 identity 决定路径边界与 P16 plan gate 的主体。
  readonly context: ToolContext;
  // Hook 或规则建议，最终仍需经过 strongestDecision 合并。
  readonly recommendations?: readonly PermissionDecision[];
  // 当前 ask 决策交给审批器时的候选结论。
  readonly proposedDecision?: PermissionDecision;
}

export class PermissionRequest {
  // 权限决策唯一消费的调用快照。
  readonly prepared: PreparedToolCall;
  // 执行身份和工作区边界，不能由工具输入伪造。
  readonly context: ToolContext;
  // 所有参与者的建议冻结副本，防止审批过程中被修改。
  readonly recommendations: readonly PermissionDecision[];
  // 仅 ask 请求存在；allow/deny 不能伪装成待审批状态。
  readonly proposedDecision: PermissionDecision | undefined;

  constructor(options: PermissionRequestOptions) {
    if (
      options.prepared.error !== undefined ||
      options.prepared.definition === undefined ||
      options.prepared.arguments === undefined
    ) {
      throw new PermissionContractError("permission request requires a valid prepared tool call");
    }
    const recommendations = options.recommendations === undefined ? [] : options.recommendations;
    if (
      !Array.isArray(recommendations) ||
      !recommendations.every((decision) => decision instanceof PermissionDecision)
    ) {
      throw new PermissionContractError("recommendations must contain PermissionDecision values");
    }
    if (
      options.proposedDecision !== undefined &&
      (!(options.proposedDecision instanceof PermissionDecision) ||
        options.proposedDecision.behavior !== "ask")
    ) {
      throw new PermissionContractError("proposedDecision must be an ask PermissionDecision");
    }
    this.prepared = options.prepared;
    this.context = options.context;
    this.recommendations = Object.freeze([...recommendations]);
    this.proposedDecision = options.proposedDecision;
    Object.freeze(this);
  }
}

export type PermissionMatcher = (request: PermissionRequest) => boolean | Promise<boolean>;

export interface PermissionRuleOptions {
  // 规则名称同时作为审计 source 和异常定位标签。
  readonly name: string;
  // 匹配后提出的候选行为，最终仍受系统硬边界约束。
  readonly behavior: PermissionBehavior;
  // 面向审计和审批器的稳定说明。
  readonly reason: string;
  // 根据完整 PermissionRequest 决定规则是否适用。
  readonly matches: PermissionMatcher;
}

export class PermissionRule {
  // 冻结规则身份，避免运行中修改 plan gate 的语义。
  readonly name: string;
  // 匹配成功时贡献的候选行为。
  readonly behavior: PermissionBehavior;
  // 匹配成功时写入决策的稳定理由。
  readonly reason: string;
  // 可异步访问协议状态的匹配器。
  readonly matches: PermissionMatcher;

  constructor(options: PermissionRuleOptions) {
    if (typeof options.name !== "string" || options.name.trim().length === 0) {
      throw new PermissionContractError("permission rule name must not be empty");
    }
    if (!isPermissionBehavior(options.behavior)) {
      throw new PermissionContractError("permission rule behavior must be a PermissionBehavior");
    }
    if (typeof options.reason !== "string" || options.reason.trim().length === 0) {
      throw new PermissionContractError("permission rule reason must not be empty");
    }
    if (typeof options.matches !== "function") {
      throw new PermissionContractError("permission rule matcher must be callable");
    }
    this.name = options.name;
    this.behavior = options.behavior;
    this.reason = options.reason;
    this.matches = options.matches;
    Object.freeze(this);
  }

  async evaluate(request: PermissionRequest): Promise<PermissionDecision | undefined> {
    // matcher 支持异步；规则自身不吞异常，策略层会把异常统一转成 deny。
    if (!(await this.matches(request))) {
      return undefined;
    }
    return new PermissionDecision(this.behavior, this.reason, this.name);
  }
}

export interface ApprovalProvider {
  // 只接收已合并的 ask 请求，必须返回显式 allow 或 deny 才能放行。
  decide(request: PermissionRequest): Promise<PermissionDecision>;
}

export interface AuditSink {
  // 记录最终决策；审计失败由策略调用边界向上暴露。
  record(request: PermissionRequest, decision: PermissionDecision): Promise<void>;
}

export interface PermissionPolicyOptions {
  // 顺序注册的业务规则，最终按行为强度合并。
  readonly rules?: readonly PermissionRule[];
  // ask 决策的人工或自动审批边界。
  readonly approval?: ApprovalProvider;
  // 最终决策的持久审计边界。
  readonly audit?: AuditSink;
  // write 工具的真实 workspace 路径校验器。
  readonly writeBoundary?: WorkspaceWriteBoundary;
}

export class PermissionPolicy {
  // 规则冻结快照；withRules 通过新实例追加，避免污染 Lead 策略。
  readonly #rules: readonly PermissionRule[];
  // 缺失时 ask fail-closed 为 deny。
  readonly #approval: ApprovalProvider | undefined;
  // 可选审计 sink，记录所有最终结论。
  readonly #audit: AuditSink | undefined;
  // 只由 write 工具使用的路径边界适配器。
  readonly #writeBoundary: WorkspaceWriteBoundary | undefined;

  constructor(options: PermissionPolicyOptions = {}) {
    const rules = options.rules === undefined ? [] : options.rules;
    if (!Array.isArray(rules) || !rules.every((rule) => rule instanceof PermissionRule)) {
      throw new PermissionContractError("rules must contain PermissionRule values");
    }
    this.#rules = Object.freeze([...rules]);
    this.#approval = options.approval;
    this.#audit = options.audit;
    this.#writeBoundary = options.writeBoundary;
  }

  // 规则组合返回新策略而不是原地修改；P16 的 plan gate 借此只追加到队友策略，不污染 Lead 策略。
  withRules(rules: readonly PermissionRule[]): PermissionPolicy {
    if (!Array.isArray(rules) || !rules.every((rule) => rule instanceof PermissionRule)) {
      throw new PermissionContractError("rules must contain PermissionRule values");
    }
    return new PermissionPolicy({
      rules: [...this.#rules, ...rules],
      ...(this.#approval === undefined ? {} : { approval: this.#approval }),
      ...(this.#audit === undefined ? {} : { audit: this.#audit }),
      ...(this.#writeBoundary === undefined ? {} : { writeBoundary: this.#writeBoundary }),
    });
  }

  async decide(request: PermissionRequest): Promise<PermissionDecision> {
    if (!(request instanceof PermissionRequest)) {
      throw new PermissionContractError("request must be a PermissionRequest");
    }
    // 系统硬边界先参与合并，后续 allow 不能覆盖它产生的 deny。
    const candidates: PermissionDecision[] = [];
    const workspaceBoundary = await this.#workspaceBoundaryDecision(request);
    if (workspaceBoundary !== undefined) {
      candidates.push(workspaceBoundary);
    }
    const shellDefault = shellDefaultDecision(request);
    if (shellDefault !== undefined) {
      candidates.push(shellDefault);
    }
    candidates.push(...request.recommendations, ...(await this.#evaluateRules(request)));

    const proposed = strongestDecision(candidates);
    let final: PermissionDecision;
    if (proposed.behavior === "ask") {
      final = await this.#resolveApproval(request, proposed);
    } else if (proposed.behavior === "passthrough") {
      final = new PermissionDecision("allow", "No permission rule blocked the request", "default");
    } else {
      final = proposed;
    }
    if (this.#audit !== undefined) {
      await this.#audit.record(request, final);
    }
    return final;
  }

  async #evaluateRules(request: PermissionRequest): Promise<readonly PermissionDecision[]> {
    // 并行求值全部规则，再按 deny > ask > allow 合并；任一规则异常按 deny fail-closed。
    const results = await Promise.all(
      this.#rules.map(async (rule) => {
        try {
          return await rule.evaluate(request);
        } catch {
          return new PermissionDecision("deny", `Permission rule failed: ${rule.name}`, rule.name);
        }
      }),
    );
    return results.filter((decision): decision is PermissionDecision => decision !== undefined);
  }

  async #resolveApproval(
    request: PermissionRequest,
    proposed: PermissionDecision,
  ): Promise<PermissionDecision> {
    if (this.#approval === undefined) {
      return implicitApprovalDenial(proposed);
    }
    const approvalRequest = new PermissionRequest({
      prepared: request.prepared,
      context: request.context,
      recommendations: request.recommendations,
      proposedDecision: proposed,
    });
    let decision: unknown;
    try {
      decision = await this.#approval.decide(approvalRequest);
    } catch {
      return new PermissionDecision("deny", "Approval provider failed; request denied", "approval");
    }
    if (!(decision instanceof PermissionDecision)) {
      return new PermissionDecision(
        "deny",
        "Approval provider returned an invalid decision",
        "approval",
      );
    }
    if (decision.behavior === "allow" || decision.behavior === "deny") {
      return decision;
    }
    return implicitApprovalDenial(proposed);
  }

  async #workspaceBoundaryDecision(
    request: PermissionRequest,
  ): Promise<PermissionDecision | undefined> {
    const definition = request.prepared.definition;
    const argumentsValue = request.prepared.arguments;
    if (definition === undefined || argumentsValue === undefined) {
      throw new PermissionContractError("permission request lost validated tool data");
    }
    if (definition.effect !== "write") {
      return undefined;
    }
    const rawPath = Reflect.get(argumentsValue as object, "path");
    if (rawPath === undefined) {
      return undefined;
    }
    if (typeof rawPath !== "string") {
      return new PermissionDecision("deny", "Write path is invalid", "workspace-boundary");
    }
    if (this.#writeBoundary === undefined) {
      return undefined;
    }
    try {
      // 真实路径解析由组合根注入的文件系统边界完成，core 不依赖具体 adapter。
      const allowed = await this.#writeBoundary.isPathWithinWorkspace(
        request.context.workspace,
        rawPath,
      );
      if (typeof allowed !== "boolean") {
        throw new PermissionContractError("write boundary must return a boolean");
      }
      return allowed
        ? undefined
        : new PermissionDecision(
            "deny",
            "Writing outside the workspace is forbidden",
            "workspace-boundary",
          );
    } catch {
      return new PermissionDecision(
        "deny",
        "Write path could not be resolved safely",
        "workspace-boundary",
      );
    }
  }
}

function strongestDecision(decisions: readonly PermissionDecision[]): PermissionDecision {
  // deny 是系统硬边界，优先级最高；ask 次之；只有没有任何参与者拒绝时才落到 allow。
  for (const behavior of ["deny", "ask", "allow"] as const) {
    const decision = decisions.find((candidate) => candidate.behavior === behavior);
    if (decision !== undefined) {
      return decision;
    }
  }
  return new PermissionDecision(
    "passthrough",
    "No permission participant made a decision",
    "default",
  );
}

function implicitApprovalDenial(proposed: PermissionDecision): PermissionDecision {
  return new PermissionDecision(
    "deny",
    `Approval was not explicitly granted: ${proposed.reason}`,
    "approval",
  );
}

function shellDefaultDecision(request: PermissionRequest): PermissionDecision | undefined {
  const definition = request.prepared.definition;
  if (definition === undefined) {
    throw new PermissionContractError("permission request lost its tool definition");
  }
  if (definition.effect !== "execute") {
    return undefined;
  }
  return new PermissionDecision("ask", "Shell execution requires approval", "shell-default");
}
